pkg/service/aws/oidc.go

```diff
@@ -0,0 +1,271 @@
+
+	parsedURL.Path = path.Join(parsedURL.Path, ".well-known", "openid-configuration")
+	openidConfigURL := parsedURL.String()
+	return openidConfigURL, nil
+}
+
+// getOIDCThumbprint will retrieve the thumbprint of the root CA for the OIDC Provider identified by the issuer URL.
+// This is done by first looking up the domain where the keys are provided, and then looking up the TLS certificate
+// chain for that domain.
+func getOIDCThumbprint(ctx context.Context, issuerURL string) (string, error) {
+
+	openidConfigURL, err := getOIDCConfigURL(issuerURL)
+	if err != nil {
+		return "", errors.Wrap(err, "getOIDCConfigURL")
+	}
+
+	jwksURL, err := getJwksURL(ctx, openidConfigURL)
+	if err != nil {
+		return "", errors.Wrap(err, "getJwksURL")
+	}
+
+	return getThumbprint(jwksURL)
+}
+
+//generateTrustPolicyDocument read the current policy document as a map, create new policy using the template stringa and convert that to map
+//finally append the new policy to Statement section of the current policy
+func generateTrustPolicyDocument(currentPolicyDoc, oidcproviderArn, oidcIssuer string) (string, error) {
+	decodedValue, err := url.QueryUnescape(currentPolicyDoc)
+	if err != nil {
+		return "", err
+	}
+
+	policyDoc := make(map[string]interface{})
+	err = json.Unmarshal([]byte(decodedValue), &policyDoc)
+	if err != nil {
+		return "", err
+	}
+
+	newPolicy := getTrustPolicyDocument(oidcproviderArn, oidcIssuer)
+
+	newDoc := make(map[string]interface{})
+	err = json.Unmarshal([]byte(newPolicy), &newDoc)
+	if err != nil {
+		return "", err
+	}
+	policyDoc["Statement"] = append(policyDoc["Statement"].([]interface{}), newDoc)
+
+	b, err := json.Marshal(policyDoc)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+// fetch OIDC from the cluster and attach it to role policy
+
+func (a *awsController) RegisterClusterOIDC(ctx context.Context, req *proto.RegisterClusterOIDCRequest) (*proto.RegisterClusterOIDCResponse, error) {
+	//get the oidc endpoint from the cluster spec
+
+	roleName := config.Get().OpenIDRole
+	region := req.Region
+	accountName := req.AccountName
+	session, err := NewSession(ctx, region, accountName)
+
+	if err != nil {
+		return nil, err
+	}
+	eksClient := session.getEksClient()
+
+	cluster, err := getClusterSpec(ctx, eksClient, req.ClusterName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get the cluster")
+	}
+
+	if *cluster.Status != eks.ClusterStatusActive {
+		a.logger.Info(ctx, "cluster is not active yet", "status", *cluster.Status)
+		return nil, fmt.Errorf("cluster is not active, curent state: %s", *cluster.Status)
+	}
+	if cluster.Identity == nil {
+		a.logger.Info(ctx, "cluster doesnt have identity", "identity", nil)
+		return nil, errors.New("cluster identity is nil")
+	}
+
+	if cluster.Identity.Oidc == nil {
+		a.logger.Info(ctx, "cluster doesnt have oidc identity", "identity.oidc", nil)
+		return nil, errors.New("cluster oidc identity is nil")
+	}
+
+	issuer := *cluster.Identity.Oidc.Issuer
```

Comment:
Can we check issuer is nonempty?

Reply:
This is not handled

Reply:
had added nil check issuer is a pointer, so if its not there it should be nil, not empty. To be on safer side, will add empty check too